File: packages/java/src/test/java/com/a2alite/AgentTest.java
package com.a2alite;

import com.a2alite.testing.AgentTestClient;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

class AgentTest {

    @Test
    void shouldCreateAgentWithNameAndDescription() {
        var agent = Agent.builder()
            .name("TestBot")
            .description("A test bot")
            .build();

        assertThat(agent.getName()).isEqualTo("TestBot");
        assertThat(agent.getDescription()).isEqualTo("A test bot");
    }

    @Test
    void shouldRegisterSkill() {
        var agent = Agent.builder()
            .name("Bot")
            .description("Test")
            .build();

        agent.skill("greet", params -> "Hello!");

        assertThat(agent.getSkills()).containsKey("greet");
    }

    @Test
    void shouldBuildAgentCard() {
        var agent = Agent.builder()
            .name("TestBot")
            .description("A test bot")
            .version("2.0.0")
            .build();

        agent.skill("greet", params -> "Hello");
        agent.skill("farewell", params -> "Goodbye");

        var card = agent.buildAgentCardJson("localhost", 9000);

        assertThat(card.get("name").asText()).isEqualTo("TestBot");
        assertThat(card.get("version").asText()).isEqualTo("2.0.0");
        assertThat(card.get("skills").size()).isEqualTo(2);
    }

    @Test
    void shouldCallSkillViaTestClient() throws Exception {
        var agent = Agent.builder()
            .name("Bot")
            .description("Test")
            .build();

        agent.skill("greet", params -> "Hello, " + params.get("name") + "!");

        var client = new AgentTestClient(agent);
        var result = client.call("greet", Map.of("name", "World"));

        assertThat(result.getData()).isEqualTo("Hello, World!");
    }

    @Test
    void shouldReturnMapResults() throws Exception {
        var agent = Agent.builder()
            .name("Bot")
            .description("Test")
            .build();

        agent.skill("info", params -> Map.of(
            "id", params.get("id"),
            "status", "active"
        ));

        var client = new AgentTestClient(agent);
        var result = client.call("info", Map.of("id", 42));
        var data = (Map<?, ?>) result.getData();

        assertThat(data.get("id")).isEqualTo(42);
        assertThat(data.get("status")).isEqualTo("active");
    }

    @Test
    void shouldListSkills() {
        var agent = Agent.builder()
            .name("Bot")
            .description("Test")
            .build();

        agent.skill("skill1", params -> "one");
        agent.skill("skill2", params -> "two");

        var client = new AgentTestClient(agent);
        var skills = client.listSkills();

        assertThat(skills).contains("skill1", "skill2");
    }

    @Test
    void shouldExecuteMiddleware() throws Exception {
        var agent = Agent.builder()
            .name("Bot")
            .description("Test")
            .build();

        agent.skill("add", params ->
            (Integer) params.get("a") + (Integer) params.get("b")
        );

        // Add middleware that logs
        final StringBuilder log = new StringBuilder();
        agent.use((ctx, next) -> {
            log.append("before:");
            Object result = next.call();
            log.append("after");
            return result;
        });

        var client = new AgentTestClient(agent);
        var result = client.call("add", Map.of("a", 2, "b", 3));

        assertThat(result.getData()).isEqualTo(5);
        assertThat(log.toString()).isEqualTo("before:after");
    }
}

File: packages/typescript/tests/agent.test.ts
import { describe, it, expect } from 'vitest';
import { Agent, AgentTestClient } from '../src/index.js';

describe('Agent', () => {
  it('should create an agent with name and description', () => {
    const agent = new Agent({ name: 'TestBot', description: 'A test bot' });
    expect(agent.name).toBe('TestBot');
    expect(agent.description).toBe('A test bot');
  });

  it('should register a skill', () => {
    const agent = new Agent({ name: 'Bot', description: 'Test' });

    agent.skill('greet', async ({ name }: { name: string }) => {
      return `Hello, ${name}!`;
    });

    const card = agent.buildAgentCard();
    expect(card.skills).toHaveLength(1);
    expect(card.skills[0].name).toBe('greet');
  });

  it('should build agent card', () => {
    const agent = new Agent({
      name: 'TestBot',
      description: 'A test bot',
      version: '2.0.0',
    });

    agent.skill('greet', async () => 'Hello');
    agent.skill('farewell', async () => 'Goodbye');

    const card = agent.buildAgentCard('localhost', 9000);

    expect(card.name).toBe('TestBot');
    expect(card.version).toBe('2.0.0');
    expect(card.url).toBe('http://localhost:9000/a2a/jsonrpc');
    expect(card.skills).toHaveLength(2);
  });
});

describe('AgentTestClient', () => {
  it('should call a skill', async () => {
    const agent = new Agent({ name: 'Bot', description: 'Test' });

    agent.skill('greet', async ({ name }: { name: string }) => {
      return `Hello, ${name}!`;
    });

    const client = new AgentTestClient(agent);
    const result = await client.call('greet', { name: 'World' });

    expect(result.data).toBe('Hello, World!');
  });

  it('should return dict results', async () => {
    const agent = new Agent({ name: 'Bot', description: 'Test' });

    agent.skill('info', async ({ id }: { id: number }) => {
      return { id, status: 'active' };
    });

    const client = new AgentTestClient(agent);
    const result = await client.call('info', { id: 42 });

    expect(result.data).toEqual({ id: 42, status: 'active' });
  });

  it('should list skills', () => {
    const agent = new Agent({ name: 'Bot', description: 'Test' });

    agent.skill('skill1', async () => 'one');
    agent.skill('skill2', async () => 'two');

    const client = new AgentTestClient(agent);
    const skills = client.listSkills();

    expect(skills).toContain('skill1');
    expect(skills).toContain('skill2');
  });

  it('should stream results', async () => {
    const agent = new Agent({ name: 'Bot', description: 'Test' });

    agent.skill('count', { streaming: true }, async function* ({ n }: { n: number }) {
      for (let i = 1; i <= n; i++) {
        yield `Count: ${i}`;
      }
    });

    const client = new AgentTestClient(agent);
    const results = await client.stream('count', { n: 3 });

    expect(results).toEqual(['Count: 1', 'Count: 2', 'Count: 3']);
  });
});
